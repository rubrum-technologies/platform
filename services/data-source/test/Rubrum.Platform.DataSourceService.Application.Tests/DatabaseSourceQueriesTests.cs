using CookieCrumble;
using Npgsql;
using Shouldly;
using Testcontainers.PostgreSql;
using Xunit;
using static Rubrum.Platform.DataSourceService.PostgresHelper;

namespace Rubrum.Platform.DataSourceService;

public class DatabaseSourceQueriesTests : DataSourceServiceApplicationGraphqlTestBase
{
    [Fact]
    public async Task SchemaDatabase_Postgresql()
    {
        var container = new PostgreSqlBuilder().Build();

        await container.StartAsync();

        await using var dataSource = NpgsqlDataSource.Create(container.GetConnectionString());

        await CreateTableUsersAsync(dataSource);
        await CreateTableDocumentsAsync(dataSource);

        await using var result = await ExecuteRequestAsync(b => b.SetDocument(
            $$"""
              query {
                  schemaDatabase(kind: POSTGRESQL, connectionString: "{{container.GetConnectionString()}}") {
                      ... on DatabaseSchemaInformation {
                          tables {
                              name
                              columns {
                                  kind
                                  name
                              }
                          }
                      }
                  }
              }
              """));

        await container.DisposeAsync().AsTask();

        result.ShouldNotBeNull();
        result.MatchSnapshot();
    }

    [Fact]
    public async Task SchemaDatabase_Postgresql_FailConnectException()
    {
        await using var result = await ExecuteRequestAsync(b => b.SetDocument(
            """
              query {
                  schemaDatabase(kind: POSTGRESQL, connectionString: "Host=127.0.0.1;Port=55200;Database=postgres;Username=postgres;Password=postgres") {
                      ... on DatabaseSchemaInformation {
                          tables {
                              name
                              columns {
                                  kind
                                  name
                              }
                          }
                      }
                      ... on FailConnectError {
                        connectionString
                        message
                      }
                      __typename
                  }
              }
              """));

        result.ShouldNotBeNull();
        result.MatchSnapshot();
    }

    [Fact]
    public async Task SchemaDatabase_Postgresql_IncorrectСonnectionStringException()
    {
        await using var result = await ExecuteRequestAsync(b => b.SetDocument(
            """
            query {
                schemaDatabase(kind: POSTGRESQL, connectionString: "some connection string") {
                    ... on DatabaseSchemaInformation {
                        tables {
                            name
                            columns {
                                kind
                                name
                            }
                        }
                    }
                    ... on IncorrectConnectionStringError {
                      incorrectConnectionString
                      message
                    }
                    __typename
                }
            }
            """));

        result.ShouldNotBeNull();
        result.MatchSnapshot();
    }
}
