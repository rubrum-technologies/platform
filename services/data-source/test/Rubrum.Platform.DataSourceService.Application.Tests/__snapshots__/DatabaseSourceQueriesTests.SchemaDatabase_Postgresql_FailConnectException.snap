{
  "data": {
    "schemaDatabase": {
      "connectionString": "Host=127.0.0.1;Port=55200;Database=postgres;Username=postgres;Password=postgres",
      "message": "Failed to connect to 127.0.0.1:55200",
      "__typename": "FailConnectError"
    }
  }
}
